### PR TITLE
Automated GCP Credentials Addon Prototype

### LINKED ISSUE #8379: GCP Auth addon
Figure out how to automatically plumb GCP credentials into minikube pods

For ease of use for anyone using GCP services in the k8s app in minikube.

### PR DESCRIPTION
**Pull request by sharifelgamal**
You can enable the addon with `minikube addons enable gcp-auth`. Credentials are retrieved from wherever GCP would normally find them, which varies platform to platform. The credentials file location can always be overridden by specifying the GOOGLE_APPLICATION_CREDENTIALS environment variable.

If a user doesn't want a specific pod to have creds injected, they can add the "gcp-auth-skip-secret" label to their pod yaml.

Fixes #8379

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1,2 +1 @@
 out/buildroot*
-
```
Comment (medyagh) [inline .dockerignore:2]: remove unrealted change

File: pkg/addons/config.go [hunk 2]
```diff
@@ -161,4 +164,9 @@ var Addons = []*Addon{
 		set:       SetBool,
 		callbacks: []setFn{enableOrDisableAddon},
 	},
+	{
+		name:      "metadata",
```
Comment (tstromberg) [inline pkg/addons/config.go:168]: This should be called 'gcp-metadata', to disambiguate it from other sources.

File: cmd/metadata-server/server.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+				Message: err.Error(),
+			},
+		}
+	}
+
+	var patch []patchOperation
+
+	// Define the volume to mount in
+	v := corev1.Volume{
+		Name: "gcp-creds",
```
Comment (tstromberg) [inline cmd/metadata-server/server.go:89]: I'm honestly a little iffy about storing vendor-specific binaries within the minikube repository, but would like to hear what other folks feel about this. Should this binary go into its own repo?

File: .dockerignore (hunk 1 above)
Comment (sharifelgamal) [reply .dockerignore:2]: there's no reason to have a newline there but sure

File: pkg/addons/metadata/enable.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	if enable {
+		return enableAddon(cfg)
+	}
+	return disableAddon(cfg)
+
+}
+
+func enableAddon(cfg *config.ClusterConfig) error {
+	// This is the default location for GCP credentials to live, it's where they're stored when gcloud login is run
+	credsPath := filepath.Join(homedir.HomeDir(), ".config", "gcloud", "application_default_credentials.json")
```
Comment (tstromberg) [inline pkg/addons/metadata/enable.go:47]: I found this function which seems more portable: https://godoc.org/golang.org/x/oauth2/google#FindDefaultCredentials

File: cmd/metadata-server/server.go (hunk 3 above)
Comment (sharifelgamal) [reply cmd/metadata-server/server.go]: Yeah this has been moved.

File: pkg/addons/metadata/enable.go (hunk 4 above)
Comment (sharifelgamal) [reply pkg/addons/metadata/enable.go:47]: I don't actually want the credentials, I just want the json file.

File: pkg/addons/config.go (hunk 2 above)
Comment (sharifelgamal) [reply pkg/addons/config.go:168]: I'm going to change it to gcp-auth for the time being.

File: pkg/addons/metadata/enable.go (hunk 4 above)
Comment (sharifelgamal) [reply pkg/addons/metadata/enable.go]: Fixed.

File: README.md [hunk 5]
```diff
@@ -66,4 +66,4 @@ minikube is a Kubernetes [#sig-cluster-lifecycle](https://github.com/kubernetes/
 
 Join our meetings:
 * [Bi-weekly office hours, Mondays @ 11am PST](https://tinyurl.com/minikube-oh)
-* [Triage Party](https://minikube.sigs.k8s.io/docs/contrib/triage/)
\ No newline at end of file
+* [Triage Party](https://minikube.sigs.k8s.io/docs/contrib/triage/)
```
Comment (medyagh) [inline README.md:69]: revert unrelated changes

File: pkg/addons/gcpauth/enable.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+		return err
+	}
+
+	err = r.Copy(f)
+	if err != nil {
+		return err
+	}
+
+	// We're currently assuming gcloud is installed and in the user's path
+	project, err := exec.Command("gcloud", "config", "get-value", "project").Output()
```
Comment (medyagh) [inline pkg/addons/gcpauth/enable.go:76]: I believe there is an environment variable that has the google project in it I suggest ,first try to get by Env variable, if it failed, 
then try this method as back up fail over plan !

File: pkg/addons/gcpauth/enable.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		return err
+	}
+
+	r, err := machine.CommandRunner(host)
+	if err != nil {
+		return err
+	}
+
+	// Clean up the files generated when enabling the addon
+	creds := assets.NewMemoryAssetTarget([]byte{}, "/tmp/google_application_credentials.json", "0444")
```
Comment (medyagh) [inline pkg/addons/gcpauth/enable.go:102]: put this path in a var in the package to be used for both enable and disable

File: deploy/addons/gcp-auth/gcp-auth-webhook.yaml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+          image: gcr.io/k8s-minikube/gcp-auth-webhook:latest
+          imagePullPolicy: IfNotPresent
+          ports:
+            - containerPort: 8443
+          volumeMounts:
+          - name: webhook-certs
+            mountPath: /etc/webhook/certs
+            readOnly: true
+          - name: gcp-project
+            mountPath: /tmp/google_cloud_project
```
Comment (medyagh) [inline deploy/addons/gcp-auth/gcp-auth-webhook.yaml:91]: I think this would be vanished, if a pod crashes and restarts, (and will loose creds) mount it to somewhere else such as /gcp-auth//google_cloud_project

File: deploy/addons/gcp-auth/gcp-auth-webhook.yaml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+- name: gcp-auth-mutate.k8s.io
+  sideEffects: None
+  admissionReviewVersions: ["v1", "v1beta1"]
+  clientConfig:
+    service:
+      name: gcp-auth
+      namespace: gcp-auth
+      path: "/mutate"
+  rules:
+  - operations: ["CREATE", "UPDATE"]
```
Comment (medyagh) [inline deploy/addons/gcp-auth/gcp-auth-webhook.yaml:141]: I suggest we have a label, that if a user explicity wanted NOT to get the secret mounted to a pod.

for example if a user deploys a pod with a label
"minikube.k8s.io/gcp-auth-do-not-mount-secret=True" then we skip mounting that.

that is useful for people who do not trust all the things they deploy to their cluster such as third party apps.

File: pkg/addons/gcpauth/enable.go (hunk 6 above)
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go:76]: yeah it's the GOOGLE_CLOUD_PROJECT env var

File: pkg/addons/config.go [hunk 10]
```diff
@@ -161,4 +164,9 @@ var Addons = []*Addon{
 		set:       SetBool,
 		callbacks: []setFn{enableOrDisableAddon},
 	},
+	{
+		name:      "gcp-auth",
+		set:       SetBool,
+		callbacks: []setFn{gcpauth.EnableOrDisable, enableOrDisableAddon},
```
[the remark this answers is not in the document]
Comment (sharifelgamal) [reply pkg/addons/config.go:170]: verifyAddonStatus hardcodes "kube-system" as the namespace to check, and this addon creates its own addon. I'm going to forego adding a verification layer until we can specify the namespace to look at

File: deploy/addons/gcp-auth/gcp-auth-webhook.yaml (hunk 8 above)
Comment (sharifelgamal) [reply deploy/addons/gcp-auth/gcp-auth-webhook.yaml]: yeah I moved it to /var/lib/minikube

File: deploy/addons/gcp-auth/gcp-auth-webhook.yaml (hunk 9 above)
Comment (sharifelgamal) [reply deploy/addons/gcp-auth/gcp-auth-webhook.yaml:147]: object selector added

File: pkg/addons/gcpauth/enable.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+	// Grab credentials from where GCP would normally look
+	ctx := context.Background()
+	creds, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return err
+	}
+
+	f := assets.NewMemoryAssetTarget(creds.JSON, "/tmp/google_application_credentials.json", "0444")
+
+	api, err := machine.NewAPIClient()
```
[the remark this answers is not in the document]
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go]: done

File: pkg/addons/config.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sharifelgamal) [reply pkg/addons/config.go]: I tweaked verifyAddonStatus to make this work

File: pkg/addons/gcpauth/enable.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+func enableAddon(cfg *config.ClusterConfig) error {
+	// Grab credentials from where GCP would normally look
+	ctx := context.Background()
+	creds, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return err
+	}
+
+	f := assets.NewMemoryAssetTarget(creds.JSON, "/tmp/google_application_credentials.json", "0444")
```
[the remark this answers is not in the document]
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go]: done

File: pkg/addons/gcpauth/enable.go (hunk 7 above)
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go]: done

File: pkg/addons/gcpauth/enable.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return err
+	}
+
+	project := assets.NewMemoryAssetTarget([]byte{}, projectPath, "0444")
+	err = r.Remove(project)
+	if err != nil {
+		return err
+	}
+
```
Comment (medyagh) [inline pkg/addons/gcpauth/enable.go:103]: after enalbing this addon, there should be an out.T

that tells the user what risks they might have

"After enabliing this addon your GCP creds will be mounted to all the created pods, if you dno't want a pod to have the secret, add this label"

File: pkg/addons/gcpauth/enable.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+	// First check if the project env var is explicitly set
+	projectEnv := os.Getenv("GOOGLE_CLOUD_PROJECT")
+	if projectEnv != "" {
+		f := assets.NewMemoryAssetTarget([]byte(projectEnv), projectPath, "0444")
+		return r.Copy(f)
+	}
+
+	// We're currently assuming gcloud is installed and in the user's path
+	project, err := exec.Command("gcloud", "config", "get-value", "project").Output()
```
[the remark this answers is not in the document]
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go:77]: that's exactly what this is, we only get to this code if the env is empty

File: pkg/addons/gcpauth/enable.go (hunk 13 above)
Comment (sharifelgamal) [reply pkg/addons/gcpauth/enable.go:104]: done

enabling the addon now looks like

```
mk addons enable gcp-auth
🔎  Verifying gcp-auth addon...
📌  Your GCP credentials will now be mounted into every pod created in the minikube cluster.
📌  If you don't want credential mounted into a specific pod, add a label with the `gcp-auth-skip-secret` key to your pod configuration.
🌟  The 'gcp-auth' addon is enabled
```

**Comment by Olshansk** [comment]
This change is 💯 💯 💯
